Add logical meter formula generator for EV power

daniel-zullo-frequenz opened the pull request:
Create a formula engine from the component graph for calculating total EV Chargers power.

Addresses #121

daniel-zullo-frequenz commented:
This is just an initial draft. I still need to test it, update/add logical meter tests and add an entry to the release notes

shsms commented on src/frequenz/sdk/timeseries/logical_meter/_formula_generators/_ev_charger_power_formula.py:
```diff
@@ -0,0 +1,56 @@
+        Returns:
+            A formula engine that calculates total EV charger power values.
+
+        Raises:
+            ComponentNotFound: if the component graph doesn't have a `GRID` component.
```
This won't happen here,  because we are not looking for a grid component while generating this formula.

(I realize the other EV current formula has this in the docstring too.  I guess my vision is just sharper when I'm reading other peoples code. 😆  )

daniel-zullo-frequenz replied:
Then I'll add a separate commit to fix that in EV current formula too

daniel-zullo-frequenz replied:
done

shsms commented:
Hi @daniel-zullo-frequenz , could you please rebase this on #180 and write tests that follow the new scheme?  Or better yet, you could wait a few days for #180 to get merged, before you start working on the tests.

daniel-zullo-frequenz commented:
> Hi @daniel-zullo-frequenz , could you please rebase this on https://github.com/frequenz-floss/frequenz-sdk-python/pull/180 and write tests that follow the new scheme? Or better yet, you could wait a few days for https://github.com/frequenz-floss/frequenz-sdk-python/pull/180 to get merged, before you start working on the tests.

Sure, it makes sense. I'll be watching the progress on #180 :eyes: and work out this patch and tests of top of it

daniel-zullo-frequenz committed:
Fix documentation in EV current formula generator

The EV current formula generator won't raise
`ComponentNotFound` because it is not looking for
a grid component while generating the formula.

Signed-off-by: Daniel Zullo <daniel.zullo@frequenz.com>

daniel-zullo-frequenz committed:
Add logical meter formula for EV power

Create a formula engine from the component graph for
calculating total EV Chargers power.

Signed-off-by: Daniel Zullo <daniel.zullo@frequenz.com>

daniel-zullo-frequenz commented on tests/timeseries/test_logical_meter.py:
```diff
@@ -328,24 +343,31 @@ async def test_formula_composition(  # pylint: disable=too-many-locals
+        engine = (
+            pv_power_recv.clone() + battery_power_recv.clone() + ev_power_recv.clone()
+        ).build("inv_power")
         inv_calc_recv = engine.new_receiver()
```
@sahas-subramanian-frequenz is it correct to include `ev_power` as part of `inv_power`?

daniel-zullo-frequenz replied:
We agreed on create a dedicated test for EV power

daniel-zullo-frequenz commented:
I have only updated existent unit tests, please let me know if you think I should add new tests instead

ela-kotulska-frequenz commented on src/frequenz/sdk/timeseries/logical_meter/_formula_generators/_ev_charger_power_formula.py:
```diff
@@ -0,0 +1,53 @@
+        Returns:
+            A formula engine that calculates total EV charger power values.
+        """
+        builder = self._get_builder("ev-power", ComponentMetricId.ACTIVE_POWER)
+        component_graph = microgrid.get().component_graph
+        ev_chargers = [
+            comp
+            for comp in component_graph.components()
+            if comp.category == ComponentCategory.EV_CHARGER
+        ]
```
@sahas-subramanian-frequenz  should we check if there is meter before ev_chargers?
And if there is a meter we could get active_power from the meter.

shsms replied:
Let's do that after we have a better idea of what all the possible layouts are.

It is all resampled data that we are operating on here,  so it should be minimal difference.

ela-kotulska-frequenz commented on tests/timeseries/test_logical_meter.py:
```diff
@@ -176,14 +176,15 @@ async def test_grid_power_2(
         assert self._equal_float_lists(results, meter_sums)
 
-    async def test_battery_and_pv_power(  # pylint: disable=too-many-locals
+    async def test_battery_pv_and_ev_power(  # pylint: disable=too-many-locals
```
It would be better to test one thing in one test.
Could you add new test for ev_power only?

daniel-zullo-frequenz commented:
Added a dedicated unit test for EV formula and an entry to the release notes

daniel-zullo-frequenz commented:
I rebased yesterday to the latest changes in the repo using github feature and now it complains about `The base branch requires all commits to be signed`. So I'm afraid I'll need to re-push it again to be able to merge it
